specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,318 @@
+                            int response = MessageBox(
+                                nullptr, message.c_str(), L"Launching External Protocol",
+                                MB_YESNOCANCEL | MB_ICONWARNING);
+                            if (response == IDYES)
+                            {
+                                std::wstring protocol_url = uri.get();
+                                SHELLEXECUTEINFO info = {sizeof(info)};
+                                info.fMask = SEE_MASK_NOASYNC;
+                                info.lpVerb = L"open";
+                                info.lpFile = protocol_url.c_str();
+                                info.nShow = SW_SHOWNORMAL;
+                                ::ShellExecuteEx(&info);
+                            }
+                        }
+                        return S_OK;
+                    };
+                    wil::com_ptr<ICoreWebView2Deferral> deferral;
+                    CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                    m_appWindow->RunAsync(
+                        [deferral, showDialog]()
+                        {
+                            showDialog();
+                            CHECK_FAILURE(deferral->Complete());
+                        });
+                    return S_OK;
+                })
+                .Get(),
+            &m_launchingExternalProtocolToken));
+    }
+}
+    
+```
+
+## .NET and WinRT
+
+```c#
+private WebView2 webView;
+void RegisterLaunchingExternalProtocolHandler() 
+{
+    webView.CoreWebView2.LaunchingExternalProtocol += (sender, args) {
+        {
+            CoreWebView2Deferral deferral = args.GetDeferral();
+            System.Threading.SynchronizationContext.Current.Post((_) =>
+            {
+                using (deferral)
+                {
+                    // Set the `Cancel` property to `TRUE`, as we will either silently launch the 
+                    // trusted app or display a custom dialog. 
+                    args.Cancel = true;
+                    if (String.Equals(args.Uri, "calculator:///", StringComparison.OrdinalIgnoreCase))
+                    {
+                        // If this matches our desired protocol, then set the
+                        // event args to cancel the event and launch the
+                        // calculator app.
+                        ProcessStartInfo info = new ProcessStartInfo
+                        {
+                            FileName = args.Uri,
+                            UseShellExecute = true
+                        };
+                        Process.Start(info);
+                    } 
+                    else
+                    {
+                        // To display a custom dialog we cancel the launch, display
+                        // a custom dialog, and then manually launch the external protocol. 
+                        string text = "Launching External Protocol";
+                        if (args.InitiatingOrigin != "")

```

Comment:
Do we promise string.Empty for the InitiatingOrigin if not applicable? Or could this be `null`, and we should use `if (!string.IsNullOrEmpty(args.InitiatingOrigin))`?

Reply:
We promise string.Empty if not applicable.

Reply:
Please make sure we document this.